File: dotnet/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.CachingProxyExample/RealWeatherService.cs
namespace ProxyDesignPattern.CachingProxyExample;

/// <summary>
/// RealSubject - The actual weather service that makes expensive API calls.
/// Each call takes time and resources (simulated with delay).
/// </summary>
public class RealWeatherService : IWeatherService
{
    public string GetWeather(string city)
    {
        Console.WriteLine($"[RealService] Fetching weather data for '{city}'...");
        Console.WriteLine("[RealService] Connecting to weather API...");
        Thread.Sleep(2000); // Simulate slow API call

        // Simulate weather data
        var conditions = new[] { "Sunny", "Cloudy", "Rainy", "Snowy", "Windy" };
        var temp = Random.Shared.Next(-10, 35);
        var condition = conditions[Random.Shared.Next(conditions.Length)];

        var result = $"{condition}, {temp}C";
        Console.WriteLine($"[RealService] Data received: {result}");
        
        return result;
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.CachingProxyExample/src/RealWeatherService.ts
import { IWeatherService } from "./IWeatherService";

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

/**
 * RealSubject - The actual weather service that makes expensive API calls.
 * Each call takes time and resources (simulated with delay).
 */
export class RealWeatherService implements IWeatherService {
  async getWeather(city: string): Promise<string> {
    console.log(`[RealService] Fetching weather data for '${city}'...`);
    console.log("[RealService] Connecting to weather API...");
    await sleep(2000); // Simulate slow API call

    const conditions = ["Sunny", "Cloudy", "Rainy", "Snowy", "Windy"];
    const temp = Math.floor(Math.random() * 45) - 10; // -10 to 35
    const condition = conditions[Math.floor(Math.random() * conditions.length)];

    const result = `${condition}, ${temp}C`;
    console.log(`[RealService] Data received: ${result}`);

    return result;
  }
}
